tests/python/pants_test/reporting/test_json_reporter.py

```diff
@@ -0,0 +1,160 @@
+    self.end_time = self.start_time + 10
+
+  def unaccounted_time(self):
+    return 0
+
+  def root(self):
+    ret = self
+    while ret.parent is not None:
+      ret = ret.parent
+    return ret
+
+  def outputs(self):
+    return {}
+
+  def outcome(self):
+    return WorkUnit.SUCCESS
+
+  def outcome_string(self, outcome):
+    return WorkUnit.outcome_string(outcome)
+
+
+class JsonReporterTest(TestBase):
+
+  def test_reporter_output(self):
+    expected = {
+      'workunits': {
+        'root_id': {
+          'name': 'root',
+          'id': 'root_id',
+          'parent_name': '',
+          'parent_id': '',
+          'labels': [
+            'IAMROOT'
+          ],
+          'cmd': '',
+          'start_time': -5419800.0,
+          'outputs': {},
+          'children': [
+            {
+              'name': 'child1',
+              'id': 'child1_id',
+              'parent_name': 'root',
+              'parent_id': 'root_id',
+              'labels': [],
+              'cmd': '',
+              'start_time': 31564800.0,
+              'outputs': {},
+              'children': [
+                {
+                  'name': 'grandchild',
+                  'id': 'grandchild_id',
+                  'parent_name': 'child1',
+                  'parent_id': 'child1_id',
+                  'labels': [
+                    'LABEL1'
+                  ],
+                  'cmd': '',
+                  'start_time': 479721600.0,
+                  'outputs': {},
+                  'children': [],
+                  'log_entries': [],
+                  'outcome': 'SUCCESS',
+                  'end_time': 479721610.0,
+                  'unaccounted_time': 0
+                }
+              ],
+              'log_entries': [],
+              'outcome': 'SUCCESS',
+              'end_time': 31564810.0,
+              'unaccounted_time': 0
+            },
+            {
+              'name': 'child2',
+              'id': 'child2_id',
+              'parent_name': 'root',
+              'parent_id': 'root_id',
+              'labels': [],
+              'cmd': '',
+              'start_time': 684140400.0,
+              'outputs': {},
+              'children': [],
+              'log_entries': [],
+              'outcome': 'SUCCESS',
+              'end_time': 684140410.0,
+              'unaccounted_time': 0
+            }
+          ],
+          'log_entries': [],
+          'outcome': 'SUCCESS',
+          'end_time': -5419790.0,
+          'unaccounted_time': 0
+        },
+      },
+      'artifact_cache_stats': []
+    }
+
+    with temporary_dir() as temp_dir:
+      run_tracker = FakeRunTracker()
+      reporter = JsonReporter(run_tracker, JsonReporter.Settings(log_level=Report.INFO, json_dir=temp_dir))
+      run_tracker.reporter = reporter
+
+      run_tracker.start()
+
+      with run_tracker.new_workunit(FakeWorkUnit(None, 'root', labels=['IAMROOT'], start_time=-5419800.0)) as root_workunit:
```

Comment:
This is good but a bit brittle and possibly over-elaborate. You have to manually create `FakeWorkUnit` instances, and keep them in sync with your expected struct.  

You can make this test simpler and yet more comprehensive, by iterating over some workunit struct (made out of dicts), creating FakeWorkUnits and generating callbacks on the fly, and then checking that you get the *same* struct back! Then you can try this on several different structs, and you don't have to keep the input and the expected output in sync.

Something like this:

```
def make_fake_workunit(wu_dict, parent_wu):
  return FakeWorkUnit(...) # Set up what you need from wu_dict and parent_wu.

def generate_callbacks(wu_dict, parent_wu=None):
  wu = make_fake_workunit(wu_dict, parent_wu)
  reporter.start_workunit(fake_wu)
  for child_dict in wu_dict.children:
    generate_callbacks(child_dict, wu_dict)
  reporter.end_workunit(fake_wu)

def check_callbacks(root_wu_dict):
  generated_wu_dict = generate_callbacks(root_wu_dict)
  self.assertDictEqual(root_wu_dict, generated_wu_dict)
```

I'm sure I got a couple of details wrong there, but does this make sense in general? 

Then it's easy to call `check_callbacks` on a variety of different structs, to test various edge cases and so on.

Reply:
PS This would make it easy to address @baroquebobcat's request for more test cases.